craft_application/commands/lifecycle.py

```diff
@@ -0,0 +1,286 @@
+
+        step_name = cast(str, step_name or self.name)
+
+        self._lifecycle_service.run(step_name=step_name, part_names=parsed_args.parts)
+
+
+class PullCommand(_LifecycleStepCommand):
+    """Command to pull parts."""
+
+    name = "pull"
+    help_msg = "Download or retrieve artifacts defined for a part"
+    overview = textwrap.dedent(
+        """
+        Download or retrieve artifacts defined for a part. If part names
+        are specified only those parts will be pulled, otherwise all parts
+        will be pulled.
+        """
+    )
+
+
+class OverlayCommand(_LifecycleStepCommand):
+    """Command to overlay parts."""
+
+    name = "overlay"
+    help_msg = "Create part layers over the base filesystem."
+    overview = textwrap.dedent(
+        """
+        Execute operations defined for each part on a layer over the base
+        filesystem, potentially modifying its contents.
+        """
+    )
+
+
+class BuildCommand(_LifecycleStepCommand):
+    """Command to build parts."""
+
+    name = "build"
+    help_msg = "Build artifacts defined for a part"
+    overview = textwrap.dedent(
+        """
+        Build artifacts defined for a part. If part names are specified only
+        those parts will be built, otherwise all parts will be built.
+        """
+    )
+
+
+class StageCommand(_LifecycleStepCommand):
+    """Command to stage parts."""
+
+    name = "stage"
+    help_msg = "Stage built artifacts into a common staging area"
+    overview = textwrap.dedent(
+        """
+        Stage built artifacts into a common staging area. If part names are
+        specified only those parts will be staged. The default is to stage
+        all parts.
+        """
+    )
+
+
+class PrimeCommand(_LifecycleStepCommand):
+    """Command to prime parts."""
+
+    name = "prime"
+    help_msg = "Prime artifacts defined for a part"
+    overview = textwrap.dedent(
+        """
+        Prepare the final payload to be packed, performing additional
+        processing and adding metadata files. If part names are specified only
+        those parts will be primed. The default is to prime all parts.
+        """
+    )
+
+    _package_service: services.PackageService
+
+    def __init__(self, config: dict[str, Any]) -> None:
+        super().__init__(config)
+        self._package_service = cast(
+            services.PackageService, config.get("package_service")
+        )
+
+    @override
+    def run(
+        self, parsed_args: "argparse.Namespace", step_name: str | None = None
+    ) -> None:
+        """Run the prime command."""
+        super().run(parsed_args, step_name=step_name)
+
+        self._package_service.write_metadata(self._lifecycle_service.prime_dir)
+
+
+class PackCommand(PrimeCommand):
+    """Command to pack the final artifact."""
+
+    run_managed = False
+
+    name = "pack"
+    help_msg = "Create the final artifact"
+    overview = textwrap.dedent(
+        """
+        Process parts and create the final artifact.
```

Comment:
Have we given some thought on how applications are going to potentially "override" this overview/help_msg? I ask this because e.g. for Snapcraft the current help for `pack` refers to snaps (which is more meaningful to the Snapcraft user), etc

Reply:
Probably just subclassing and generating an app-specific command group, sadly.